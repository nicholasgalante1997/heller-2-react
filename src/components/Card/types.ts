import { HTMLProps } from 'react';

export interface CardProps extends Omit<HTMLProps<HTMLDivElement>, 'children' | 'media' | 'action'> {
  media: string | JSX.Element;
  mediaAlt?: string;
  title: string;
  subtitle?: string;
  description: string;
  action: ButtonAction | LinkAction;
  scale?: number;
}

export interface MicroCardProps extends Omit<CardProps, 'media' | 'mediaAlt' | 'action'> {
    action: LinkAction;
}

interface ButtonAction {
  type: 'button';
  onChange: (event: React.FormEvent<HTMLButtonElement>) => void;
  label: string;
}

interface LinkAction {
  type: 'link';
  href: string;
  target?: HTMLProps<HTMLAnchorElement>['target'];
  label: string;
}
