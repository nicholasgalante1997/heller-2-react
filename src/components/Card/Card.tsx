import React, { memo } from 'react';
import classNames from 'classnames';

import { Button } from '../Button';
import { Link } from '../Link';

import { CardProps } from './types';
import { Heading } from '../Heading';
import { Body } from '../Body';

function Card({
  action,
  description,
  media,
  mediaAlt,
  scale,
  subtitle,
  title,
  className = '',
  style = {},
  ...rest
}: CardProps) {
  const mediaIsJsx = typeof media !== 'string';

  return (
    <div
      {...rest}
      style={
        {
          ...style,
          '--hlr-card-default-height-scale-factor': scale,
          '--hlr-card-default-width-scale-factor': scale
        } as any
      }
      className={classNames(className, 'hlr-card')}
    >
      <div className="hlr-card-image-container">
        {mediaIsJsx ? media : <img src={media} alt={mediaAlt} />}
      </div>
      <div className="hlr-card-body-container">
        <Heading as="h1">{title}</Heading>
        {subtitle && <Body className="hlr-card-body-subtitle" as="span">{subtitle}</Body>}
        <Body className="hlr-card-body-description" as="p">{description}</Body>
      </div>
      <div className="hlr-card-action-container">
        {action.type === 'button' && <Button onChange={action.onChange}>{action.label}</Button>}
        {action.type === 'link' && <Link href={action.href} target={action.target}>{action.label}</Link>}
      </div>
    </div>
  );
}

export default memo(Card as React.FC<CardProps>);
