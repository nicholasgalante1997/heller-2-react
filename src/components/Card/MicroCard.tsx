import React, { memo } from 'react';
import classNames from 'classnames';

import { Link } from '../Link';

import { MicroCardProps } from './types';
import { Heading } from '../Heading';
import { Body } from '../Body';

function MicroCard({
  action,
  description,
  scale,
  subtitle,
  title,
  className = '',
  style = {},
  ...rest
}: MicroCardProps) {
  return (
    <div
      {...rest}
      style={
        {
          ...style,
          '--hlr-card-default-height-scale-factor': scale,
          '--hlr-card-default-width-scale-factor': scale
        } as any
      }
      className={classNames(className, 'hlr-micro-card')}
    >
      <div className="hlr-card-body-container">
        <Heading as="h1">{title}</Heading>
        {subtitle && <Body className="hlr-card-body-subtitle" as="span">{subtitle}</Body>}
        <Body className="hlr-card-body-description" as="p">{description}</Body>
      </div>
      <div className="hlr-card-action-container">
        <Link href={action.href} target={action.target}>{action.label}</Link>
      </div>
    </div>
  );
}

export default memo(MicroCard as React.FC<MicroCardProps>);