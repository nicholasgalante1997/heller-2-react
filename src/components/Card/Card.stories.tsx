import { Meta, StoryObj } from '@storybook/react';
import React from 'react';
import Card from './Card';
import MicroCard from './MicroCard';

const meta: Meta<typeof Card> = {
    title: 'components/Card',
    component: Card
};
export default meta;

export const CardPG: StoryObj<typeof Card> = {
    args: {
        title: "Air Max 90",
        media: "./air-max.jpeg",
        mediaAlt: "Red Nike Air Max Shoes, side profile.",
        subtitle: "Popular",
        description: "A popular favorite is back for a limited time. Maximum order number 25. Ships August 2024.",
        action: {
            type: "button",
            label: "Add to Cart",
            onChange: () => alert('Added to cart!')
        }
    },
    render: (args) => <Card {...args} />
};

export const MicroCardPG: StoryObj<typeof MicroCard> = {
    args: {
        title: "Air Max 90",
        subtitle: "Popular",
        description: "A popular favorite is back for a limited time. Maximum order number 25. Ships August 2024.",
        action: {
            type: "link",
            label: "Add to Cart",
            href: "#",
            target: "_self"
        }
    },
    render: (args) => <MicroCard {...args} />
};