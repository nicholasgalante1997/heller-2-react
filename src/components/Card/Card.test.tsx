import React from 'react';
import { render, screen } from '@testing-library/react';
import Card from './Card';
import MicroCard from './MicroCard';

describe('Card', () => {
  describe('MicroCard', () => {
    test('Snapshot', () => {
        const jsx = <MicroCard action={{href: "#", label: 'Read', type: 'link', target: '_self'}} description='Stub' title='Stub' />;
        const { container } = render(jsx);
        expect(container).toMatchSnapshot();
    });
  });
  describe('Card', () => {
    test('Snapshot', () => {
        const jsx = <Card media="stub" mediaAlt="alt-stub" action={{href: "#", label: 'Read', type: 'link', target: '_self'}} description='Stub' title='Stub' />;
        const { container } = render(jsx);
        expect(container).toMatchSnapshot();
    });
  });
});
