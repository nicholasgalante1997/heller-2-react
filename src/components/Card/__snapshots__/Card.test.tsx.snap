// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Card Snapshot 1`] = `
<div>
  <div
    class="hlr-card"
  >
    <div
      class="hlr-card-image-container"
    >
      <img
        alt="alt-stub"
        src="stub"
      />
    </div>
    <div
      class="hlr-card-body-container"
    >
      <h1
        class="h-heading-100"
      >
        Stub
      </h1>
      <p
        class="h-body hlr-card-body-description"
      >
        Stub
      </p>
    </div>
    <div
      class="hlr-card-action-container"
    >
      <a
        class="h-link"
        href="#"
        target="_self"
      >
        Read
      </a>
    </div>
  </div>
</div>
`;

exports[`Card MicroCard Snapshot 1`] = `
<div>
  <div
    class="hlr-micro-card"
  >
    <div
      class="hlr-card-body-container"
    >
      <h1
        class="h-heading-100"
      >
        Stub
      </h1>
      <p
        class="h-body hlr-card-body-description"
      >
        Stub
      </p>
    </div>
    <div
      class="hlr-card-action-container"
    >
      <a
        class="h-link"
        href="#"
        target="_self"
      >
        Read
      </a>
    </div>
  </div>
</div>
`;
